timApp/tide_support/routes.py

```diff
@@ -0,0 +1,399 @@
+                )
+            )
+
+    return tasks
+
+
+def ide_task_by_id(
+        user: User,
+        ide_task_id: str,
+        ide_task_tag: str,
+        doc_id: int = None,
+        doc_path: str = None,
+) -> json:
+    """
+    Get TIDE-task from the document by document id and paragraph id
+    :param ide_task_tag: Tag that identifies the TIDE-task
+    :param ide_task_id:  TIDE-task id
+    :param user: Logged-in user
+    :param doc_id:  Document id
+    :param doc_path:  Document path
+    :return: JSON with TIDE-task ide-files, task info and task id
+    """
+    if doc_id is None and doc_path is None:
+        return []
+
+    if doc_path is None:
+        doc = DocInfo.find_by_id(item_id=doc_id)
+    else:
+        doc = DocInfo.find_by_path(path=doc_path.lower())
+
+    if doc is None:
+        return []
+
+    user_ctx = UserContext.from_one_user(u=user)
+
+    pars = doc.document.get_paragraphs()
+
+    if pars is None:
+        return []
+
+    if len(pars) == 1:
+        if pars[0].attrs.get(ide_task_tag) == ide_task_id:
+            return user_plugin_data(doc=doc, par=pars[0], user_ctx=user_ctx)
+        else:
+            return []
+
+    task = []
+
+    for p in pars:
+        if p.attrs.get(ide_task_tag) == ide_task_id:
+            task.append(user_plugin_data(doc=doc, par=p, user_ctx=user_ctx))
+
+    # # In case multiple files with same names
+    # if len(task) > 1:
+    #     return task
+
+    return task
+
+
+def user_plugin_data(
+        doc: DocInfo,
+        par,
+        user_ctx: UserContext,
+        plugin: Plugin = None,
+        ide_task_id: str = None,
+) -> json:
+    """
+    Get the TIDE-task information from the plugin
+    :param ide_task_id:  TIDE-task id
+    :param plugin: Tim plugin
+    :param doc: TIM document
+    :param par: Paragraph from the document
+    :param user_ctx: User context
+    :return: JSON with TIDE-task ide-files, task info and task id
+    """
+
+    view_ctx = default_view_ctx
+
+    if plugin is None:
+        plugin = Plugin.from_paragraph(par, view_ctx, user_ctx)
+
+    # Plugin render options
+    plugin_opts = PluginRenderOptions(
+        do_lazy=False,
+        user_print=False,
+        preview=view_ctx.preview,
+        target_format=PrintFormat.JSON,
+        output_format=PluginOutputFormat.HTML,
+        user_ctx=user_ctx,
+        review=False,
+        wraptype=PluginWrap.Nothing,
+        viewmode=view_ctx.viewmode,
+    )
+    plugin.set_render_options(None, plugin_opts)
+    res = render_plugin_multi(doc.document.get_settings(), "csPlugin", [plugin])
+    plugin_htmls = json.loads(res)
+    plugin_html = BeautifulSoup(plugin_htmls[0], features="lxml")
+    element = plugin_html.select_one("cs-runner")
+    plugin_json = json.loads(base64.b64decode(element.attrs["json"]).decode("utf-8"))
+
+    task_info: TIDETaskInfo = TIDETaskInfoSchema.load(plugin.values, unknown=EXCLUDE)
+
+    task_id = plugin_json["taskID"]
```

Comment:
Mielellään tyypitetty versio

```suggestion
    task_id: TaskId | None = plugin.task_id
```

huomaa, että pluginilla periaatteessa voi olla olemattakin IDtä. Silloin se ei ole validi tehtävä eikä sitä pitäisi hyväksyä.

`TaskId`-tyypistä saa merkkijonon kutsumalla `task_id.doc_task`

Reply:
korjattu